File: python/lsst/meas/modelfit/psf.py
#!/usr/bin/env python
#
# LSST Data Management System
# Copyright 2008-2014 LSST Corporation.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.
#

import lsst.pex.config
import lsst.meas.base
from . import modelfitLib


class ShapeletPsfApproxConfig(lsst.pex.config.Config):
    models = lsst.pex.config.ConfigDictField(
        keytype=str,
        itemtype=modelfitLib.PsfFitterConfig,
        doc="a dictionary of models that can be used to fit the PSF",
        default={} # populated in setDefaults; can't do it on a single line here
        )
    sequence = lsst.pex.config.ListField(
        dtype=str,
        doc="a sequence of model names indicating which models should be fit, and their order",
        default=["DoubleShapelet"]
        )

    def setDefaults(self):
        super(ShapeletPsfApproxConfig, self).setDefaults()
        self.models["SingleGaussian"] = modelfitLib.PsfFitterConfig()
        self.models["SingleGaussian"].inner.order = -1
        self.models["SingleGaussian"].primary.order = 0
        self.models["SingleGaussian"].wings.order = -1
        self.models["SingleGaussian"].outer.order = -1
        self.models["DoubleGaussian"] = modelfitLib.PsfFitterConfig()
        self.models["DoubleGaussian"].inner.order = -1
        self.models["DoubleGaussian"].primary.order = 0
        self.models["DoubleGaussian"].wings.order = 0
        self.models["DoubleGaussian"].outer.order = -1
        self.models["DoubleShapelet"] = modelfitLib.PsfFitterConfig()
        self.models["DoubleShapelet"].inner.order = -1
        self.models["DoubleShapelet"].primary.order = 2
        self.models["DoubleShapelet"].wings.order = 1
        self.models["DoubleShapelet"].outer.order = -1
        self.models["Full"] = modelfitLib.PsfFitterConfig()
        self.models["Full"].inner.order = 0
        self.models["Full"].primary.order = 4
        self.models["Full"].wings.order = 4
        self.models["Full"].outer.order = 0

    def validate(self):
        super(ShapeletPsfApproxConfig, self).validate()
        if len(self.sequence) < 1:
            raise ValueError("sequence must have at least one element")
        for m in self.sequence:
            if m not in self.models:
                raise KeyError("All elements in sequence must be keys in models dict")

class ShapeletPsfApproxMixin(object):
    """Mixin base class for fitting shapelet approximations to the PSF model

    This class does almost all of the work for its two derived classes, ShapeletPsfApproxSingleFramePlugin
    and ShapeletPsfApproxForcedPlugin, which simply adapt it to the slightly different interfaces for
    single-frame and forced measurement.  It in turn delegates its work to the C++ PsfFitter class;
    it holds sequence of these corresponding to different models (generally with increasing complexity).
    Each PsfFitter starts with the result of the previous one as an input, using PsfFitter::adapt to
    hopefully allow these previous fits to reduce the time spent on the next one.
    """

    def __init__(self, config, name, schema):
        """Initialize the plugin, creating a sequence of PsfFitter instances to do the fitting and
        MultiShapeletFunctionKey instances to save the results to a record.
        """
        self.sequence = []
        for m in config.sequence:
            fitter = modelfitLib.PsfFitter(config.models[m].makeControl())
            modelKey = fitter.addModelFields(schema, schema.join(name, m))
            flagKey = schema.addField(schema.join(name, m, "flag"), type="Flag",
                                      doc="General failure flag set if anything goes wrong.")
            nIterKey = schema.addField(schema.join(name, m, "nIterations"), type=int,
                                       doc="Number of E-M steps taken by the fitter.")
            self.sequence.append((fitter, name, modelKey, flagKey, nIterKey))

    def measure(self, measRecord, exposure):
        """Fit the configured sequence of models the given Exposure's Psf, as evaluated at
        measRecord.getCentroid(), then save the results to measRecord.
        """
        if not exposure.hasPsf():
            raise lsst.meas.base.FatalAlgorithmError("ShapeletPsfApprox requires Exposure to have a Psf")
        psf = exposure.getPsf()
        psfImage = psf.computeKernelImage(measRecord.getCentroid())
        psfShape = psf.computeShape(measRecord.getCentroid())
        model = None
        lastFitter = None
        # Fit the first element in the sequence, using the PSFs moments to initialize the parameters
        # For every other element in the fitting sequence, use the previous fit to initialize the parameters
        for fitter, name, modelKey, flagKey, nIterKey in self.sequence:
            try:
                if model is None:
                    model = fitter.makeInitial(psfShape)
                else:
                    model = fitter.adapt(model, lastFitter)
                lastFitter = fitter
                nIterations = fitter.apply(model, psfImage)
                measRecord.set(modelKey, model)
                measRecord.set(nIterKey, nIterations)
                if nIterations == fitter.getMaxIterations():
                    measRecord.set(flagKey, True)
            except Exception:
                # Since this is an unexpected failure mode, we throw now and do not proceed to the next fit.
                measRecord.set(flagKey, True)
                raise

    # This plugin doesn't need to set a flag in fail(), because it should have been
    # done already by the try/except in measure().
    def fail(self, measRecord, error=None):
        pass

class ShapeletPsfApproxSingleFrameConfig(lsst.meas.base.SingleFramePluginConfig, ShapeletPsfApproxConfig):

    def setDefaults(self):
        lsst.meas.base.SingleFramePluginConfig.setDefaults(self)
        ShapeletPsfApproxConfig.setDefaults(self)

@lsst.meas.base.register("modelfit_ShapeletPsfApprox")
class ShapeletPsfApproxSingleFramePlugin(lsst.meas.base.SingleFramePlugin, ShapeletPsfApproxMixin):
    """Minimal subclass of ShapeletPsfApproxMixin to conform to the single-frame measurement API.

    This class simply provides __init__ and measure methods that matched the SingleFramePlugin signatures
    and delegate to the ShapeletPsfApproxMixin's implmeentations.
    """
    ConfigClass = ShapeletPsfApproxSingleFrameConfig

    @staticmethod
    def getExecutionOrder():
        return 1.0

    def __init__(self, config, name, schema, metadata):
        ShapeletPsfApproxMixin.__init__(self, config, name, schema)
        lsst.meas.base.SingleFramePlugin.__init__(self, config, name, schema, metadata)

    def measure(self, measRecord, exposure):
        ShapeletPsfApproxMixin.measure(self, measRecord, exposure)

    def fail(self, measRecord, error=None):
        ShapeletPsfApproxMixin.fail(self, measRecord, error)

class ShapeletPsfApproxForcedConfig(lsst.meas.base.ForcedPluginConfig, ShapeletPsfApproxConfig):

    def setDefaults(self):
        lsst.meas.base.ForcedPluginConfig.setDefaults(self)
        ShapeletPsfApproxConfig.setDefaults(self)

@lsst.meas.base.register("modelfit_ShapeletPsfApprox")
class ShapeletPsfApproxForcedPlugin(lsst.meas.base.ForcedPlugin, ShapeletPsfApproxMixin):
    """Minimal subclass of ShapeletPsfApproxMixin to conform to the forced measurement API.

    This class simply provides __init__ and measure methods that matched the ForcedPlugin signatures
    and delegate to the ShapeletPsfApproxMixin's implmeentations.
    """
    ConfigClass = ShapeletPsfApproxForcedConfig

    @staticmethod
    def getExecutionOrder():
        return 1.0

    def __init__(self, config, name, schemaMapper, metadata):
        ShapeletPsfApproxMixin.__init__(self, config, name, schemaMapper.editOutputSchema())
        lsst.meas.base.ForcedPlugin.__init__(self, config, name, schemaMapper, metadata)

    def measure(self, measRecord, exposure, refRecord, refWcs):
        ShapeletPsfApproxMixin.measure(self, measRecord, exposure)

    def fail(self, measRecord, error=None):
        ShapeletPsfApproxMixin.fail(self, measRecord, error)

File: python/lsst/meas/modelfit/psf/psf.cc
// -*- lsst-c++ -*-
/*
 * LSST Data Management System
 * Copyright 2008-2013 LSST Corporation.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */

#include "pybind11/pybind11.h"
#include "lsst/cpputils/python.h"

#include "lsst/pex/config/python.h"

#include "lsst/meas/modelfit/DoubleShapeletPsfApprox.h"
#include "lsst/meas/modelfit/GeneralPsfFitter.h"

namespace py = pybind11;
using namespace pybind11::literals;

namespace lsst {
namespace meas {
namespace modelfit {
namespace {

void declareDoubleShapelet(lsst::cpputils::python::WrapperCollection &wrappers) {
    using Control = DoubleShapeletPsfApproxControl;
    using Algorithm = DoubleShapeletPsfApproxAlgorithm;

    using PyControl = py::class_<Control>;
    using PyAlgorithm = py::class_<Algorithm, meas::base::SimpleAlgorithm>;

    static auto clsControl =
            wrappers.wrapType(PyControl(wrappers.module, "DoubleShapeletPsfApproxControl"), [](auto &mod, auto &cls) {
                cls.def(py::init<>());
                LSST_DECLARE_CONTROL_FIELD(cls, Control, innerOrder);
                LSST_DECLARE_CONTROL_FIELD(cls, Control, outerOrder);
                LSST_DECLARE_CONTROL_FIELD(cls, Control, radiusRatio);
                LSST_DECLARE_CONTROL_FIELD(cls, Control, peakRatio);
                LSST_DECLARE_CONTROL_FIELD(cls, Control, minRadius);
                LSST_DECLARE_CONTROL_FIELD(cls, Control, minRadiusDiff);
                LSST_DECLARE_CONTROL_FIELD(cls, Control, maxRadiusBoxFraction);
                LSST_DECLARE_NESTED_CONTROL_FIELD(cls, Control, optimizer);
            });

    wrappers.wrapType(PyAlgorithm(wrappers.module, "DoubleShapeletPsfApproxAlgorithm"), [](auto &mod, auto &cls) {
        // wrap anonymous enum values as ints because we'll need to use them as ints
        cls.attr("Control") = clsControl;
        cls.attr("FAILURE") = py::cast(Algorithm::FAILURE);
        cls.attr("INVALID_POINT_FOR_PSF") = py::cast(Algorithm::INVALID_POINT_FOR_PSF);
        cls.attr("INVALID_MOMENTS") = py::cast(Algorithm::INVALID_MOMENTS);
        cls.attr("MAX_ITERATIONS") = py::cast(Algorithm::MAX_ITERATIONS);

        cls.def(py::init<Control const &, std::string const &, afw::table::Schema &>(), "ctrl"_a,
                         "name"_a, "schema"_a);
        cls.def_static("initializeResult", &Algorithm::initializeResult, "ctrl"_a);
        cls.def_static("fitMoments", &Algorithm::fitMoments, "result"_a, "ctrl"_a, "psfImage"_a);
        cls.def_static("makeObjective", &Algorithm::makeObjective, "moments"_a, "ctrl"_a, "psfImage"_a);
        cls.def_static("fitProfile", &Algorithm::fitProfile, "result"_a, "ctrl"_a, "psfImage"_a);
        cls.def_static("fitShapelets", &Algorithm::fitShapelets, "result"_a, "ctrl"_a, "psfImage"_a);
        cls.def("measure", &Algorithm::measure, "measRecord"_a, "exposure"_a);
        cls.def("fail", &Algorithm::fail, "measRecord"_a, "error"_a = nullptr);
    });
}

void declareGeneral(lsst::cpputils::python::WrapperCollection &wrappers) {
    using ComponentControl = GeneralPsfFitterComponentControl;
    using Control = GeneralPsfFitterControl;
    using Fitter = GeneralPsfFitter;
    using Algorithm = GeneralPsfFitterAlgorithm;

    using PyComponentControl = py::class_<ComponentControl>;
    using PyControl = py::class_<Control>;
    using PyFitter = py::class_<Fitter>;
    using PyAlgorithm = py::class_<Algorithm, Fitter>;

    wrappers.wrapType(PyComponentControl(wrappers.module, "GeneralPsfFitterComponentControl"), [](auto &mod, auto &cls) {
        cls.def(py::init<int, double>(), "order"_a = 0, "radius"_a = 1.0);
        LSST_DECLARE_CONTROL_FIELD(cls, ComponentControl, order);
        LSST_DECLARE_CONTROL_FIELD(cls, ComponentControl, positionPriorSigma);
        LSST_DECLARE_CONTROL_FIELD(cls ,ComponentControl, ellipticityPriorSigma);
        LSST_DECLARE_CONTROL_FIELD(cls, ComponentControl, radiusFactor);
        LSST_DECLARE_CONTROL_FIELD(cls, ComponentControl, radiusPriorSigma);
    });

    static auto clsControl =
            wrappers.wrapType(PyControl(wrappers.module, "GeneralPsfFitterControl"), [](auto &mod, auto &cls) {
                cls.def(py::init<>());
                LSST_DECLARE_NESTED_CONTROL_FIELD(cls, Control, inner);
                LSST_DECLARE_NESTED_CONTROL_FIELD(cls, Control, primary);
                LSST_DECLARE_NESTED_CONTROL_FIELD(cls, Control, wings);
                LSST_DECLARE_NESTED_CONTROL_FIELD(cls, Control, outer);
                LSST_DECLARE_NESTED_CONTROL_FIELD(cls, Control, optimizer);
                LSST_DECLARE_CONTROL_FIELD(cls, Control, defaultNoiseSigma);
            });

    wrappers.wrapType(PyFitter(wrappers.module, "GeneralPsfFitter"), [](auto &mod, auto &cls) {
        cls.def(py::init<Control const &>(), "ctrl"_a);
        cls.def("addFields", &Fitter::addFields, "schema"_a, "prefix"_a);
        cls.def("getModel", &Fitter::getModel);
        cls.def("getPrior", &Fitter::getPrior);
        cls.def("adapt", &Fitter::adapt, "previousFit"_a, "previousModel"_a);
        // We use lambdas here because the C++ signature has an optional int*
        // argument not relevant for Python (which confuses pybind11).
        cls.def("apply", [](Fitter const &self, afw::image::Image<Pixel> const &image,
                                  afw::geom::ellipses::Quadrupole const &moments,
                                  Scalar noiseSigma) { return self.apply(image, moments, noiseSigma); },
                      "image"_a, "moments"_a, "noiseSigma"_a = -1);
        cls.def("apply", [](Fitter const &self, afw::image::Image<double> const &image,
                                  afw::geom::ellipses::Quadrupole const &moments,
                                  Scalar noiseSigma) { return self.apply(image, moments, noiseSigma); },
                      "image"_a, "moments"_a, "noiseSigma"_a = -1);
        cls.def("apply", [](Fitter const &self, afw::image::Image<Pixel> const &image,
                                  shapelet::MultiShapeletFunction const &initial,
                                  Scalar noiseSigma) { return self.apply(image, initial, noiseSigma); },
                      "image"_a, "initial"_a, "noiseSigma"_a = -1);
        cls.def("apply", [](Fitter const &self, afw::image::Image<double> const &image,
                                  shapelet::MultiShapeletFunction const &initial,
                                  Scalar noiseSigma) { return self.apply(image, initial, noiseSigma); },
                      "image"_a, "initial"_a, "noiseSigma"_a = -1);
    });

    wrappers.wrapType(PyAlgorithm(wrappers.module, "GeneralPsfFitterAlgorithm"), [](auto &mod, auto &cls) {
        cls.attr("Control") = clsControl;
        cls.attr("FAILURE") = py::cast(Algorithm::FAILURE);
        cls.attr("MAX_INNER_ITERATIONS") = py::cast(Algorithm::MAX_INNER_ITERATIONS);
        cls.attr("MAX_OUTER_ITERATIONS") = py::cast(Algorithm::MAX_OUTER_ITERATIONS);
        cls.attr("EXCEPTION") = py::cast(Algorithm::EXCEPTION);
        cls.attr("CONTAINS_NAN") = py::cast(Algorithm::CONTAINS_NAN);
        cls.def(py::init<Control const &, afw::table::Schema &, std::string const &>(), "ctrl"_a,
                         "schema"_a, "prefix"_a);
        cls.def("getKey", &Algorithm::getKey);
        cls.def("measure",
                         (void (Algorithm::*)(afw::table::SourceRecord &, afw::image::Image<double> const &,
                                              shapelet::MultiShapeletFunction const &) const) &
                                 Algorithm::measure,
                         "measRecord"_a, "image"_a, "initial"_a);
        cls.def("measure",
                         (void (Algorithm::*)(afw::table::SourceRecord &, afw::image::Image<double> const &,
                                              afw::geom::ellipses::Quadrupole const &) const) &
                                 Algorithm::measure,
                         "measRecord"_a, "image"_a, "moments"_a);
        cls.def("fail", &Algorithm::fail, "measRecord"_a, "error"_a = nullptr);
    });
    // MultiShapeletPsfLikelihood intentionally not exposed to Python.
}

}  // namespace

void wrapPsf(lsst::cpputils::python::WrapperCollection &wrappers) {
    declareDoubleShapelet(wrappers);
    declareGeneral(wrappers);
}

}  // namespace modelfit
}  // namespace emeas
}  // namespace lsst:
